### PR TITLE
Terraform 0.12

### PR DESCRIPTION
**Pull request by mars**
Upgrade this provider to support [Terraform 0.12](https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md).

Thanks to @appilon for getting this started, I've rebased to current Terraform & Heroku provider master branches.

### CODE REVIEW

**Comment by mars** [comment]
I've been locally using the Heroku provider with Terraform master/0.12-dev and it's working nicely 👍

**Comment by mars** [comment]
I've verified that [Terraform 0.12 Beta 1](https://www.hashicorp.com/blog/announcing-terraform-0-1-2-beta1) is working with the [0.12 dev snapshot](http://terraform-0.12.0-dev-snapshots.s3-website-us-west-2.amazonaws.com/terraform-provider-heroku/) of this provider in my experimental [Heroku Terraforming app](https://github.com/mars/terraforming-app).

I'm not sure how those dev snapshots are built though, so this is not direct feedback about this pull request, but an acknowledgement that master of this provider is known to work with Terraform 0.12 🤷‍♂️

**Comment by mars** [comment]
Closing to reopen a fresh upgrade branch.

**Comment by mars** [comment]
Running [Hashicorp CI on this PR](https://ci-oss.hashicorp.engineering/viewType.html?buildTypeId=HerokuProvider_ProviderHeroku&branch_HerokuProvider=179&tab=buildTypeStatusDiv) to see what broke 😝

**Comment by davidji99** [comment]
👀

**Comment by mars** [comment]
yes @davidji, HashiCorp CI runs the full acceptance test suite using a Heroku-internal Enterprise Team.

**Comment by mars** [comment]
Hashicorp CI was successful ✅

I've verified locally that this Heroku Provider branch works with both Terraform 0.11 & 0.12 ✅

So, I think this is ready to release 🏁

**Comment by davidji99** [comment]
I tried this branch out locally with existing `heroku_app` + configs managed by terraform. I noticed a two things:
- Structure of the state file changed per the upgrade to `terraform v0.12`.
- As `heroku_app.config_vars`, `heroku_app.sensitive_config_vars`, and `heroku_app.all_config_vars` schema type changed, `terraform plan` detected an update to do (`Plan: 0 to add, 1 to change, 0 to destroy.`) around the config vars. I applied said `plan` and checked the config vars and did not notice any edits to them. There was also no rolling restarts on the dynos either.

Other than that, I don't think I noticed anything else.

File: heroku/resource_heroku_app.go [hunk 1]
```diff
@@ -158,20 +158,17 @@ func resourceHerokuApp() *schema.Resource {
...
-				Elem: &schema.Schema{
-					Type: schema.TypeMap,
-				},
 			},
 
 			"sensitive_config_vars": {
-				Type:     schema.TypeList,
+				Type:     schema.TypeMap,
 				Optional: true,
 				Computed: true,
```
Comment (davidji99) [inline heroku/resource_heroku_app.go]: Can you add a `Sensitive: true,` here? Without it, the output shows the plaintext value.

Comment (mars) [reply heroku/resource_heroku_app.go]: 👍 Done

**Comment by mars** [comment]
Thank you for the practical review @davidji99 ☺️

The "1 to change" state churn you encountered is not surprising to me, given the way the resource schema has shifted to direct `TypeMap`'s. Glad to hear there was no side-effect!

I'm going to let this chill here until early next week (Monday or Tuesday), at which point I will merge and release, unless further feedback indicates more work to be done.

**Comment by ryansdwilson** [comment]
@mars thank you for your work on this. I'm getting started with Heroku terraform and eager to try it out.

**Pull request merged**